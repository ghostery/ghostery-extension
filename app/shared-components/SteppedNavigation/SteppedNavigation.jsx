/**
 * Stepped Navigation Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';

/**
 * A Functional React component for a Stepped Navigation
 * @return {JSX} JSX for rendering a Stepped Navigation
 * @memberof SharedComponents
 */
const SteppedNavigation = (props) => {
	const {
		totalSteps,
		activeIndex,
		hrefPrev,
		hrefNext,
		hrefDone,
		textPrev,
		textNext,
		textDone,
	} = props;

	const circles = [];
	for (let i = 1; i <= totalSteps; i++) {
		circles.push((i === activeIndex) ?
			<a className="active" key={`nav-${i}`} /> :	// eslint-disable-line jsx-a11y/anchor-has-content
			<NavLink to={`${i}`} key={`nav-${i}`} />		// ^ Use <a> without an href prop to prevent a call to history with the same URL
		);
	}

	return (
		<div className="row align-center">
			<div className="columns small-12">
				{hrefDone && (
					<NavLink to={hrefDone} className="SteppedNavigation__exit flex-container align-middle">
						{textDone && (
							<span className="SteppedNavigation__exitText">{textDone}</span>
						)}
						<span className="SteppedNavigation__exitIcon" />
					</NavLink>
				)}
				<div className="SteppedNavigation flex-container">
					<div className="flex-child-grow flex-container flex-dir-row-reverse">
						{(hrefPrev && textPrev) ? (
							<div className="SteppedNavigation__buttonContainer flex-container align-center-middle">
								<NavLink className="button hollow secondary" to={hrefPrev}>
									{textPrev}
								</NavLink>
							</div>
						) : (
							<div className="SteppedNavigation__buttonContainer" />
						)}
					</div>
					<div className="SteppedNavigation__circles flex-container align-center-middle">
						{circles}
					</div>
					<div className="flex-child-grow">
						{(hrefNext && textNext) ? (
							<div className="SteppedNavigation__buttonContainer flex-container align-center-middle">
								<NavLink className="button success" to={hrefNext}>
									{textNext}
								</NavLink>
							</div>
						) : (
							<div className="SteppedNavigation__buttonContainer" />
						)}
					</div>
				</div>
			</div>
		</div>
	);
};

// PropTypes ensure we pass required props of the correct type
SteppedNavigation.propTypes = {
	totalSteps: PropTypes.number.isRequired,
	activeIndex: PropTypes.number.isRequired,
	hrefPrev: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
	hrefNext: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
	hrefDone: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
	textPrev: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
	textNext: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
	textDone: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
	]).isRequired,
};

export default SteppedNavigation;
