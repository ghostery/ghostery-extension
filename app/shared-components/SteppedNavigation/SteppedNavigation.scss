/**
 * Stepped Navigation Sass
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

// Foundation Overrides
.SteppedNavigation .button {
	margin: 0;
	min-width: 150px;
	max-width: 250px;
}

// Exit Navigation Link
.SteppedNavigation__exit {
	position: absolute;
	top: 0;
	right: 0;
	margin: 18px 33px;
	padding: 11px 13px;
	border-radius: 3px;
}
.SteppedNavigation__exitText {
	font-size: 11px;
	color: #4a4a4a;
	margin-right: 15px;
}
.SteppedNavigation__exitIcon {
	height: 15px;
	width: 15px;
	margin: 0 auto;
	background-repeat: no-repeat;
	background-position: center center;
	background-image: buildIconX(#4a4a4a);
}
.SteppedNavigation__exit:hover {
	background-color: #efefef;
}

// Bottom Navigation Circles and Buttons
.SteppedNavigation {
	margin: 60px 0 45px;
}
.SteppedNavigation__buttonContainer {
	width: 250px;
	margin: 0 15px;
}
.SteppedNavigation__circles a {
	display: inline-block;
	position: relative;
	height: 10px;
	width: 10px;
	margin: 0 10px;
	border-radius: 50%;
	background-color: #cccccc;
	border: 1px solid #a4a4a4;
}
.SteppedNavigation__circles a.active::after {
	content: '';
	display: block;
	position: absolute;
	top: -6px;
	bottom: -6px;
	left: -6px;
	right: -6px;
	border-radius: 50%;
	border: 2px solid purple;
}
