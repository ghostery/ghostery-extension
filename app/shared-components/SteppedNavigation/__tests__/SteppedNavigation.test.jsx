/**
 * Stepped Navigation Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router';
import SteppedNavigation from '../SteppedNavigation';

describe('app/shared-components/SteppedNavigation', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('stepped navigation is rendered correctly on first step', () => {
			const initialState = {
				totalSteps: 6,
				activeIndex: 1,
				hrefPrev: false,
				hrefNext: '/test/2',
				hrefDone: '/',
				textPrev: false,
				textNext: 'Next',
				textDone: 'Exit',
			};
			const component = renderer.create(
				<MemoryRouter>
					<SteppedNavigation {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('stepped navigation is rendered correctly on second step', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 2,
				hrefPrev: '/test/1',
				hrefNext: '/test/3',
				hrefDone: '/more-examples',
				textPrev: 'Previous',
				textNext: 'Foreward',
				textDone: 'Leave',
			};
			const component = renderer.create(
				<MemoryRouter>
					<SteppedNavigation {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('stepped navigation is rendered correctly on last step', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 3,
				hrefPrev: '/test/2',
				hrefNext: false,
				hrefDone: false,
				textPrev: 'Back',
				textNext: 'Not Shown',
				textDone: 'Also Not Shown',
			};

			const component = renderer.create(
				<MemoryRouter>
					<SteppedNavigation {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('More Snapshot tests with react-test-renderer, but for edge cases', () => {
		test('edge case where activeIndex beyond totalSteps', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 4,
				hrefPrev: '/test/1',
				hrefNext: '/',
				hrefDone: false,
				textPrev: 'Beginning',
				textNext: 'Exit',
				textDone: 'Not Shown',
			};
			const component = renderer.create(
				<MemoryRouter>
					<SteppedNavigation {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});

		test('edge case where hrefDone is set but textDone is not', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 1,
				hrefPrev: false,
				hrefNext: '/step/2',
				hrefDone: '/',
				textPrev: 'Not Shown',
				textNext: 'Next',
				textDone: false,
			};
			const component = renderer.create(
				<MemoryRouter>
					<SteppedNavigation {...initialState} />
				</MemoryRouter>
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('the happy path of the component', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 1,
				hrefPrev: false,
				hrefNext: '/test/2',
				hrefDone: '/',
				textPrev: false,
				textNext: 'Next',
				textDone: 'Exit',
			};

			const component = shallow(<SteppedNavigation {...initialState} />);
			expect(component.find('.SteppedNavigation__exit').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(1);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(1);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(1);

			component.setProps({
				activeIndex: 2,
				hrefPrev: '/test/1',
				hrefNext: '/test/3',
				textPrev: 'Back',
			});
			expect(component.find('.SteppedNavigation__exit').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(1);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(1);

			component.setProps({
				activeIndex: 3,
				hrefPrev: '/test/2',
				hrefNext: '/',
				hrefDone: false,
				textNext: 'Done',
				textDone: false,
			});
			expect(component.find('.SteppedNavigation__exit').length).toBe(0);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(0);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(0);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(1);
		});

		test('the edge cases of the component', () => {
			const initialState = {
				totalSteps: 3,
				activeIndex: 1,
				hrefPrev: false,
				hrefNext: '/test/2',
				hrefDone: '/',
				textPrev: false,
				textNext: 'Next',
				textDone: 'Exit',
			};

			// Happy State
			const component = shallow(<SteppedNavigation {...initialState} />);
			expect(component.find('.SteppedNavigation__exit').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(1);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(1);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(1);

			// Edge Case: activeIndex greater than the number of totalSteps
			component.setProps({
				activeIndex: 4,
				hrefPrev: '/test/1',
				hrefNext: '/',
				textPrev: 'Beginning',
				textNext: 'Exit',
			});
			expect(component.find('.SteppedNavigation__exit').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(1);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(3);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(0);

			// Edge Case: textDone is false
			component.setProps({ textDone: false });
			expect(component.find('.SteppedNavigation__exit').length).toBe(1);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(0);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(1);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(3);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(0);

			// Edge Case: hrefDone and textDone are false
			component.setProps({ hrefDone: false });
			expect(component.find('.SteppedNavigation__exit').length).toBe(0);
			expect(component.find('.SteppedNavigation__exitText').length).toBe(0);
			expect(component.find('.SteppedNavigation__exitIcon').length).toBe(0);
			expect(component.find('.SteppedNavigation__buttonContainer').length).toBe(2);
			expect(component.find('.SteppedNavigation__buttonContainer .button').length).toBe(2);
			expect(component.find('.SteppedNavigation__circles NavLink').length).toBe(3);
			expect(component.find('.SteppedNavigation__circles a.active').length).toBe(0);
		});
	});
});
