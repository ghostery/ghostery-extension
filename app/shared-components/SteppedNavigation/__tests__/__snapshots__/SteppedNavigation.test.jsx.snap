// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/SteppedNavigation More Snapshot tests with react-test-renderer, but for edge cases edge case where activeIndex beyond totalSteps 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12"
  >
    <div
      className="SteppedNavigation flex-container"
    >
      <div
        className="flex-child-grow flex-container flex-dir-row-reverse"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button hollow secondary"
            href="/test/1"
            onClick={[Function]}
            style={undefined}
          >
            Beginning
          </a>
        </div>
      </div>
      <div
        className="SteppedNavigation__circles flex-container align-center-middle"
      >
        <a
          aria-current={null}
          className={undefined}
          href="/1"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/2"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/3"
          onClick={[Function]}
          style={undefined}
        />
      </div>
      <div
        className="flex-child-grow"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current="page"
            className="button success active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Exit
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/shared-components/SteppedNavigation More Snapshot tests with react-test-renderer, but for edge cases edge case where hrefDone is set but textDone is not 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12"
  >
    <a
      aria-current="page"
      className="SteppedNavigation__exit flex-container align-middle active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      <span
        className="SteppedNavigation__exitIcon"
      />
    </a>
    <div
      className="SteppedNavigation flex-container"
    >
      <div
        className="flex-child-grow flex-container flex-dir-row-reverse"
      >
        <div
          className="SteppedNavigation__buttonContainer"
        />
      </div>
      <div
        className="SteppedNavigation__circles flex-container align-center-middle"
      >
        <a
          className="active"
        />
        <a
          aria-current={null}
          className={undefined}
          href="/2"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/3"
          onClick={[Function]}
          style={undefined}
        />
      </div>
      <div
        className="flex-child-grow"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button success"
            href="/step/2"
            onClick={[Function]}
            style={undefined}
          >
            Next
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/shared-components/SteppedNavigation Snapshot tests with react-test-renderer stepped navigation is rendered correctly on first step 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12"
  >
    <a
      aria-current="page"
      className="SteppedNavigation__exit flex-container align-middle active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      <span
        className="SteppedNavigation__exitText"
      >
        Exit
      </span>
      <span
        className="SteppedNavigation__exitIcon"
      />
    </a>
    <div
      className="SteppedNavigation flex-container"
    >
      <div
        className="flex-child-grow flex-container flex-dir-row-reverse"
      >
        <div
          className="SteppedNavigation__buttonContainer"
        />
      </div>
      <div
        className="SteppedNavigation__circles flex-container align-center-middle"
      >
        <a
          className="active"
        />
        <a
          aria-current={null}
          className={undefined}
          href="/2"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/3"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/4"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/5"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/6"
          onClick={[Function]}
          style={undefined}
        />
      </div>
      <div
        className="flex-child-grow"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button success"
            href="/test/2"
            onClick={[Function]}
            style={undefined}
          >
            Next
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`app/shared-components/SteppedNavigation Snapshot tests with react-test-renderer stepped navigation is rendered correctly on last step 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12"
  >
    <div
      className="SteppedNavigation flex-container"
    >
      <div
        className="flex-child-grow flex-container flex-dir-row-reverse"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button hollow secondary"
            href="/test/2"
            onClick={[Function]}
            style={undefined}
          >
            Back
          </a>
        </div>
      </div>
      <div
        className="SteppedNavigation__circles flex-container align-center-middle"
      >
        <a
          aria-current={null}
          className={undefined}
          href="/1"
          onClick={[Function]}
          style={undefined}
        />
        <a
          aria-current={null}
          className={undefined}
          href="/2"
          onClick={[Function]}
          style={undefined}
        />
        <a
          className="active"
        />
      </div>
      <div
        className="flex-child-grow"
      >
        <div
          className="SteppedNavigation__buttonContainer"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`app/shared-components/SteppedNavigation Snapshot tests with react-test-renderer stepped navigation is rendered correctly on second step 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12"
  >
    <a
      aria-current={null}
      className="SteppedNavigation__exit flex-container align-middle"
      href="/more-examples"
      onClick={[Function]}
      style={undefined}
    >
      <span
        className="SteppedNavigation__exitText"
      >
        Leave
      </span>
      <span
        className="SteppedNavigation__exitIcon"
      />
    </a>
    <div
      className="SteppedNavigation flex-container"
    >
      <div
        className="flex-child-grow flex-container flex-dir-row-reverse"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button hollow secondary"
            href="/test/1"
            onClick={[Function]}
            style={undefined}
          >
            Previous
          </a>
        </div>
      </div>
      <div
        className="SteppedNavigation__circles flex-container align-center-middle"
      >
        <a
          aria-current={null}
          className={undefined}
          href="/1"
          onClick={[Function]}
          style={undefined}
        />
        <a
          className="active"
        />
        <a
          aria-current={null}
          className={undefined}
          href="/3"
          onClick={[Function]}
          style={undefined}
        />
      </div>
      <div
        className="flex-child-grow"
      >
        <div
          className="SteppedNavigation__buttonContainer flex-container align-center-middle"
        >
          <a
            aria-current={null}
            className="button success"
            href="/test/3"
            onClick={[Function]}
            style={undefined}
          >
            Foreward
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
